cub/device/dispatch/dispatch_reduce.cuh

```diff
@@ -55,898 +55,1291 @@ CUB_NAMESPACE_BEGIN
+
+  /**
+   * @brief Invocation
+   *
+   * @tparam ActivePolicyT
+   *   Umbrella policy active for the target device
+   *
+   * @tparam DeviceSegmentedReduceKernelT
+   *   Function type of cub::DeviceSegmentedReduceKernel
+   *
+   * @param[in] segmented_reduce_kernel
+   *   Kernel function pointer to parameterization of 
+   *   cub::DeviceSegmentedReduceKernel
+   */
+  template <typename ActivePolicyT, typename DeviceSegmentedReduceKernelT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t
+  InvokePasses(DeviceSegmentedReduceKernelT segmented_reduce_kernel)
+  {
+    cudaError error = cudaSuccess;
+
+    do
     {
-        cudaError error = cudaSuccess;
-        do
-        {
-            // Return if the caller is simply requesting the size of the storage allocation
-            if (d_temp_storage == NULL)
-            {
-                temp_storage_bytes = 1;
-                return cudaSuccess;
-            }
-
-            // Init kernel configuration
-            KernelConfig segmented_reduce_config;
-            if (CubDebug(error = segmented_reduce_config.Init<typename ActivePolicyT::SegmentedReducePolicy>(segmented_reduce_kernel))) break;
-
-            // Log device_reduce_sweep_kernel configuration
-            #ifdef CUB_DETAIL_DEBUG_ENABLE_LOG
-            _CubLog("Invoking SegmentedDeviceReduceKernel<<<%d, %d, 0, %lld>>>(), %d items per thread, %d SM occupancy\n",
-                num_segments,
-                ActivePolicyT::SegmentedReducePolicy::BLOCK_THREADS,
-                (long long) stream,
-                ActivePolicyT::SegmentedReducePolicy::ITEMS_PER_THREAD,
-                segmented_reduce_config.sm_occupancy);
-            #endif
-
-            // Invoke DeviceReduceKernel
-            THRUST_NS_QUALIFIER::cuda_cub::launcher::triple_chevron(
-                num_segments,
-                ActivePolicyT::SegmentedReducePolicy::BLOCK_THREADS, 0, stream
-            ).doit(segmented_reduce_kernel,
-                d_in,
-                d_out,
-                d_begin_offsets,
-                d_end_offsets,
-                num_segments,
-                reduction_op,
-                init);
-
-            // Check for failure to launch
-            if (CubDebug(error = cudaPeekAtLastError()))
-            {
-                break;
-            }
-
-            // Sync the stream if specified to flush runtime errors
-            error = detail::DebugSyncStream(stream);
-            if (CubDebug(error))
-            {
-              break;
-            }
-        }
-        while (0);
-
-        return error;
-    }
-
-
-    /// Invocation
-    template <typename ActivePolicyT>
-    CUB_RUNTIME_FUNCTION __forceinline__
-    cudaError_t Invoke()
+      // Return if the caller is simply requesting the size of the storage
+      // allocation
+      if (d_temp_storage == NULL)
+      {
+        temp_storage_bytes = 1;
+        return cudaSuccess;
+      }
```

Comment:
Ditto move in front of the loop

Reply:
There's no loop :smile: See my comments to the similar observations above.